src/k8s_upgrade.py

```diff
@@ -0,0 +1,421 @@
+
+        If upgrade is not in progress, set partition to 0. (If a unit receives a stop event, it may
+        raise the partition even if an upgrade is not in progress.)
+
+        Automatically upgrades next unit if all upgraded units are healthy—except if only one unit
+        has upgraded (need manual user confirmation [via Juju action] to upgrade next unit)
+
+        Handle Juju action to:
+        - confirm first upgraded unit is healthy and resume upgrade
+        - force upgrade of next unit if 1 or more upgraded units are unhealthy
+        """
+
+        units = self._sorted_units
+
+        partition_ = self._determine_partition(
+            units,
+            action_event,
+        )
+        logger.debug(f"{self._partition=}, {partition_=}")
+        # Only lower the partition—do not raise it.
+        # If this method is called during the action event and then called during another event a
+        # few seconds later, `determine_partition()` could return a lower number during the action
+        # and then a higher number a few seconds later.
+        # This can cause the unit to hang.
+        # Example: If partition is lowered to 1, unit 1 begins to upgrade, and partition is set to
+        # 2 right away, the unit/Juju agent will hang
+        # Details: https://chat.charmhub.io/charmhub/pl/on8rd538ufn4idgod139skkbfr
+        # This does not address the situation where another unit > 1 restarts and sets the
+        # partition during the `stop` event, but that is unlikely to occur in the small time window
+        # that causes the unit to hang.
+        if partition_ < self._partition:
+            self._partition = partition_
+            logger.debug(
+                f"Lowered partition to {partition_} {action_event=} {self.in_progress=}"
+            )
+        if action_event:
+            assert len(units) >= 2
+            if self._partition > unit_number(units[1]):
+                message = "Highest number unit is unhealthy. Refresh will not resume."
+                logger.debug(f"Resume refresh event failed: {message}")
+                action_event.fail(message)
+                return
+            # If a unit was unhealthy and the upgrade was forced, only the next unit will
+            # upgrade. As long as 1 or more units are unhealthy, the upgrade will need to be
+            # forced for each unit.
+
+            # Include "Attempting to" because (on Kubernetes) we only control the partition,
+            # not which units upgrade. Kubernetes may not upgrade a unit even if the partition
+            # allows it (e.g. if the charm container of a higher unit is not ready). This is
+            # also applicable `if not force`, but is unlikely to happen since all units are
+            # healthy `if not force`.
+            message = f"Attempting to refresh unit {self._partition}."
+            action_event.set_results({"result": message})
+            logger.debug(f"Resume refresh succeeded: {message}")
```

Comment:
@Mehdi-Bendriss This is different in mongos because we have only one action so no `force` argument to handle.